### PR TITLE
PR: Introduce support for assertNotEquals for primitive data types

### PR DESCRIPTION
**Pull request by shirishp**
## Overview

Introduce support for `assertNotEquals` for primitive data types short, byte, int, long, char, float and double.

Issue #1607

Notes:
* Please review the JavaDocs in junit-jupiter-api/src/main/java/org/junit/jupiter/api/Assertions.java
* Please review implementations and tests especially of float and double with delta
* `assertNotEquals(boolean, boolean)` is implemented, but corresponding `assertEquals(boolean, boolean)` is not present. IMHO, it should be done as a separate issue and PR. I am willing to work on PR for that.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java [hunk 1]
```diff
@@ -29,6 +31,166 @@ private AssertNotEquals() {
...
+		if (unexpected == actual) {
+			failEqual(actual, message);
+		}
+	}
+
+	static void assertNotEquals(boolean unexpected, boolean actual, Supplier<String> messageSupplier) {
+		if (unexpected == actual) {
+			failEqual(actual, nullSafeGet(messageSupplier));
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: That was my initial (undocumented) response as well.

I would suggest one simply use `assertTrue()` and `assertFalse()` for `boolean` primitives.

I don't really seen a need to add `(boolean, boolean)` signatures at this point in time.

However, if the community displays a significant interest in those variants over time, we could add them at a later date.

[the remark this answers is not in the document]
Comment (lessthanoptimal) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: Use case that I could think of would be when comparing the results of two implementations. For complex algorithms I might have an easy to understand O(N^4) implementation for non production use and a much faster but complex O(N^2) implementation. The end result is something like this:

`assertEquals(naive.doSomething(data),` smart.doSomething(data)) 

Well I guess that's for `assertEquals(boolean,boolean)` and not `assertNotEquals(boolean,boolean)`...

[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: Please keep in mind that we already have `assertEquals(Object, Object)` and `assertNotEquals(Object, Object)`.

Thus, if you code something that results in `assertEquals(true, false)` or `assertNotEquals(true, false)`, the compiler simply replaces those with `assertEquals(Boolean.valueOf(true), Boolean.valueOf(false))` and `assertNotEquals(Boolean.valueOf(true), Boolean.valueOf(false))`, respectively.

So it's not that `(boolean, boolean)` is not supported. Rather, it's simply an issue of auto-boxing, and I can personally live with that.

[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: FWIW, JUnit 4 also does not have `assertEquals(boolean, boolean)` or `assertNotEquals(boolean, boolean)`.

That doesn't mean JUnit Jupiter should never introduce them. But... it just means people seem to have gotten by fine with JUnit 4 without those variants. 😉

[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: Oddly enough, JUnit 3 **did** have `assertEquals(boolean, boolean)`.

But... that delegated to:

```java
assertEquals(null, Byte.valueOf(expected), Byte.valueOf(actual));
```

🤣

[the remark this answers is not in the document]
Comment (shirishp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: Seems like we are agreeing on not having `assertNotEquals(boolean, boolean)`.
I will remove it from the PR.

[the remark this answers is not in the document]
Comment (sbrannen) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertNotEquals.java]: I think that's safe to say for this particular PR.

**Comment by shirishp** [comment]
@sbrannen @marcphilipp @lessthanoptimal 
Please review the changes. The latest changes include:
* Remove `assertNotEquals(boolean, boolean)`
* Add `@since` and `@API` annotations to methods
* Changes suggested by @lessthanoptimal except comment about tolerance, because I could not figure out how that would work with `assertNotEquals`

**Comment by sbrannen** [comment]
Thanks for making the changes.

LGTM! 👍

**Comment by sbrannen** [comment]
@junit-team/junit-lambda, FYI: a manual, local merge is _in progress_.

**Comment by sbrannen** [comment]
This PR has been merged into `master` in 0f62c6b261044009542620b6fc645217fe793ccb.

Thanks for your contribution!

**Comment by lessthanoptimal** [comment]
thanks everyone for quickly and thoughtfully adding this feature request into junit! @shirishp

**Pull request closed without merging**